### PR TITLE
feat: add support for all subquery types

### PR DESCRIPTION
**Pull request by EpsilonPrime**

### CODE REVIEW

File: src/substrait/textplan/SymbolTablePrinter.cpp [hunk 1]
```diff
@@ -671,50 +806,293 @@ void SymbolTablePrinter::addInputsToRelation(
...
+            symbolInfos,
+            expression->mutable_cast()->mutable_input(),
+            consumedPipelines);
+      }
+      break;
+    case ::substrait::proto::Expression::kSubquery:
+      // Handled below.
+      break;
+    case ::substrait::proto::Expression::kNested:
+      // TODO -- Implement nested expressions.
```
Comment (westonpace) [inline src/substrait/textplan/SymbolTablePrinter.cpp:975]: Does this silently fail, e.g. should we have a SUBSTRAIT_FAIL here?

File: src/substrait/textplan/SymbolTable.h [hunk 2]
```diff
@@ -145,8 +148,13 @@ class SymbolTable {
       const std::any& subtype,
       const std::any& blob);
 
-  // Changes the location for a specified existing symbol.
-  void updateLocation(const SymbolInfo& symbol, const Location& location);
+  // Changes the permanent location for a specified existing symbol.
+  void addPermanentLocation(const SymbolInfo& symbol, const Location& location);
```
Comment (westonpace) [inline src/substrait/textplan/SymbolTable.h:152]: Maybe expand this comment slightly to explain how "permanent" is different from source?

**Review by westonpace — APPROVED** [review]
Let's just add support for ANY in a follow-up and then I think we are good.  Thanks for working through the difficult cases here.

Some minor questions/comments for cleanup if you're interested.

File: src/substrait/textplan/parser/grammar/SubstraitPlanParser.g4 [hunk 3]
```diff
@@ -87,9 +87,17 @@ relation_detail
...
 expression
    : id LEFTPAREN (expression COMMA?)* RIGHTPAREN (ARROW literal_complex_type)? # expressionFunctionUse
-   | constant                                      # expressionConstant
-   | column_name                                   # expressionColumn
-   | expression AS literal_complex_type            # expressionCast
+   | constant                                               # expressionConstant
+   | column_name                                            # expressionColumn
+   | expression AS literal_complex_type                     # expressionCast
+   | SUBQUERY relation_ref                                  # expressionScalarSubquery
+   | expression_list IN SUBQUERY relation_ref               # expressionInPredicateSubquery
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/grammar/SubstraitPlanParser.g4:94]: https://github.com/substrait-io/substrait/issues/597

File: src/substrait/textplan/parser/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+There are four phases that the parser goes through to parse a textplan into its
+internal representation (stored inside a ```SymbolTable`````).
+
+* ```SubstraitPlanVisitor```
+   * Finds the relations, processes functions, data files, and schemas. 
+* ```SubstraitPlanPipelineVisitor```
+   * Stores the connections between the relations as determined by the pipelines 
+     section.
+* ```SubstraitPlanRelationVisitor```
+   *
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/README.md:12]: Done

File: src/substrait/textplan/parser/SubstraitPlanSubqueryRelationVisitor.h [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+class NamedStruct;
+class Type;
+class Type_Struct;
+} // namespace substrait::proto
+
+namespace io::substrait::textplan {
+
+class RelationData;
+
+class SubstraitPlanSubqueryRelationVisitor : public SubstraitPlanTypeVisitor {
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/SubstraitPlanSubqueryRelationVisitor.h:23]: Done

File: src/substrait/textplan/parser/SubstraitPlanVisitor.cpp [hunk 6]
```diff
@@ -305,25 +305,19 @@ std::any SubstraitPlanVisitor::visitRelation_type(
...
-    return RelationType::kFetch;
-  } else if (id == "aggregate") {
-    return RelationType::kAggregate;
-  } else if (id == "sort") {
-    return RelationType::kSort;
-  } else if (id == "filter") {
-    return RelationType::kFilter;
-  } else if (id == "set") {
-    return RelationType::kSet;
+  std::unordered_map<std::string, RelationType> relationTypeMap = {
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/SubstraitPlanVisitor.cpp:308]: Done

File: src/substrait/textplan/parser/grammar/SubstraitPlanLexer.g4 [hunk 7]
```diff
@@ -46,6 +46,14 @@ COUNT: 'COUNT';
 TYPE: 'TYPE';
 EMIT: 'EMIT';
 
+SUBQUERY: 'SUBQUERY';
+EXISTS: 'EXISTS';
+UNIQUE: 'UNIQUE';
+IN: 'IN';
+OR: 'OR';
+ALL: 'ALL';
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/grammar/SubstraitPlanLexer.g4:54]: Pushing to the next PR to resolve this new issue:
https://github.com/substrait-io/substrait-cpp/issues/92

File: src/substrait/textplan/parser/grammar/SubstraitPlanParser.g4 [hunk 8]
```diff
@@ -87,9 +87,17 @@ relation_detail
...
+   | (UNIQUE|EXISTS) IN SUBQUERY relation_ref (OR SUBQUERY relation_ref)*       # expressionSetPredicateSubquery
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/grammar/SubstraitPlanParser.g4:95]: fixed

File: src/substrait/textplan/converter/PlanPrinterVisitor.cpp [hunk 9]
```diff
@@ -185,6 +189,160 @@ std::string PlanPrinterVisitor::lookupFunctionReference(
...
+    case ::substrait::proto::
+        Expression_Subquery_SetComparison_ComparisonOp_COMPARISON_OP_UNSPECIFIED:
+    case ::substrait::proto::
+        Expression_Subquery_SetComparison_ComparisonOp_Expression_Subquery_SetComparison_ComparisonOp_INT_MIN_SENTINEL_DO_NOT_USE_:
+    case ::substrait::proto::
+        Expression_Subquery_SetComparison_ComparisonOp_Expression_Subquery_SetComparison_ComparisonOp_INT_MAX_SENTINEL_DO_NOT_USE_:
+      errorListener_->addError("Did not recognize the subquery comparison.");
+      return std::string("UNSUPPORTED SUBQUERY");
+  }
+  result << "ALL SUBQUERY ";
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/PlanPrinterVisitor.cpp:327]: Pushed this to the next PR as mentioned earlier.

File: src/substrait/textplan/SymbolTable.h (hunk 2 above)
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTable.h:152]: Done

File: src/substrait/textplan/SymbolTablePrinter.cpp (hunk 1 above)
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:975]: Verified that this error is caught in PlanPrinterVisitor.

**Pull request merged**